perceval/client.py

```diff
@@ -0,0 +1,185 @@
+    version = '0.1'
+
+    MIN_RATE_LIMIT = 10
+    MAX_RATE_LIMIT = 500
+
+    DEFAULT_SLEEP_TIME = 1
+    MAX_RETRIES = 5
+
+    MAX_RETRIES_ON_CONNECT = 5
+    MAX_RETRIES_ON_READ = 5
+    MAX_RETRIES_ON_REDIRECT = 5
+    MAX_RETRIES_ON_STATUS = 5
+
+    RATE_LIMIT_HEADER = "X-RateLimit-Remaining"
+    RATE_LIMIT_RESET_HEADER = "X-RateLimit-Reset"
+
+    DEFAULT_RETRY_AFTER_STATUS_CODES = [503, 413, 429]
+
+    GET = "GET"
+    POST = "POST"
+
+    def __init__(self, base_url, sleep_for_rate=False, min_rate_to_sleep=MIN_RATE_LIMIT,
+                 max_retries=MAX_RETRIES, default_sleep_time=DEFAULT_SLEEP_TIME):
+        self.base_url = base_url
+        self.rate_limit = None
+        self.rate_limit_reset_ts = None
+        self.session = None
+
+        self.sleep_for_rate = sleep_for_rate
+
+        if min_rate_to_sleep > self.MAX_RATE_LIMIT:
+            msg = "Minimum rate to sleep value exceeded (%d)."
+            msg += "High values might cause the client to sleep forever."
+            msg += "Reset to %d."
+            self.min_rate_to_sleep = self.MAX_RATE_LIMIT
+            logger.warning(msg, min_rate_to_sleep, self.MAX_RATE_LIMIT)
+        else:
+            self.min_rate_to_sleep = min_rate_to_sleep
+
+        self.max_retries = max_retries
+        self.default_sleep_time = default_sleep_time
+
+    def create_http_session(self, headers=None):
+        self.session = requests.Session()
+
+        if headers:
+            self.session.headers.update(headers)
+
+    def close_http_session(self, session):
+        session.close()
+
+    def set_retry_after_status_codes(self, status_codes=None):
+        if status_codes:
+            Retry.RETRY_AFTER_STATUS_CODES = frozenset(status_codes)
+        else:
+            Retry.RETRY_AFTER_STATUS_CODES = self.DEFAULT_RETRY_AFTER_STATUS_CODES
+
+    def init_api_token(self, path, payload=None, headers=None, method=GET):
+        session = self.session or requests.Session()
+        response = self.__send_request(session, path, payload, headers, method)
+
+        self.rate_limit = int(response.headers[self.RATE_LIMIT_HEADER])
+        self.rate_limit_reset_ts = int(response.headers[self.RATE_LIMIT_RESET_HEADER])
+
+        if not self.session:
+            self.close_http_session(session)
+
+    def fetch(self, url, payload=None, headers=None, method=GET,
+              stream=False,
+              retry_on_status_codes=None,
+              retry_after_status_code=None):
+        yield self._fetch(url, payload, headers, method, stream,
+                          retry_on_status_codes,
+                          retry_after_status_code)
+
+    def _fetch(self, url, payload=None, headers=None, method=GET,
+               stream=False,
+               retry_on_status_codes=None,
+               retry_after_status_codes=None):
+
+        self.set_retry_after_status_codes(retry_after_status_codes)
+        self.sleep_for_rate_limit()
+
+        session = self.session or requests.Session()
```

Comment:
Why doesn't the client use always a session?

I would have seen this code in a different method, maybe as part of the initialization of the instance, to call it just once.

Reply:
From the [doc](http://docs.python-requests.org/en/latest/user/advanced/#session-objects), I read that "Session object allows you to persist certain parameters across requests". I guessed that we didn't need it and it was fine to create a session per get request.

Reply:
I see, but I think the docs make reference to parameters needed by the server that won't be harmful, like in Bugzilla with the session cookies. The server may define other cookies but in my opinion, it won't be a problem. And it will keep the configuration for retrying without having to define it each time a requests is sent.